File: Scala/src/main/scala/synapses/Mathematics.scala
package synapses

object Mathematics {

  def dotProduct(xs: LazyList[Double], ys: LazyList[Double]): Double =
    xs.zip(ys)
      .map { case (x, y) => x * y }
      .sum

  def euclideanDistance(xs: LazyList[Double], ys: LazyList[Double]): Double = {
    val s = xs
      .zip(ys)
      .map { case (x, y) => (x - y) * (x - y) }
      .sum
    Math.sqrt(s)
  }

  def rootMeanSquareError(y_hats: LazyList[LazyList[Double]],
                          ys: LazyList[LazyList[Double]])
  : Double = {
    val (n, s) = y_hats
      .zip(ys)
      .map { case (y_hat, y) =>
        val d = euclideanDistance(y_hat, y)
        d * d
      }
      .foldLeft((0, 0.0)) { case ((accN, accS), sd) =>
        (accN + 1, accS + sd)
      }
    Math.sqrt(s / n)
  }

}

File: implementation/Python/synapses_py/model/mathematics.py
import math

from functional.pipeline import Sequence


def dot_product(xs: Sequence, ys: Sequence) -> float:
    return xs \
        .zip(ys) \
        .map(lambda t: t[0] * t[1]) \
        .sum()


def euclidean_distance(xs: Sequence, ys: Sequence) -> float:
    s = xs \
        .zip(ys) \
        .map(lambda t: (t[0] - t[1]) * (t[0] - t[1])) \
        .sum()
    return math.sqrt(s)


def root_mean_square_error_d(y_hat: Sequence, y: Sequence) -> float:
    d = euclidean_distance(y_hat, y)
    return d * d


def root_mean_square_error(y_hats_with_ys: Sequence) -> float:
    (n, s) = y_hats_with_ys \
        .map(lambda t: root_mean_square_error_d(t[0], t[1])) \
        .fold_left((0, 0.0), lambda acc, x: (acc[0] + 1, acc[1] + x))
    return math.sqrt(s / n)
